### ISSUE TITLE
Widening cast from Subscriber to Observer changes behavior of subscription chain

### ISSUE THREAD

**Issue by mttkay** [ReactiveX/RxJava#2599]
So this might be a debatable one, but I thought it's worth pointing out because it really leads to surprising and (I think) unexpected behavior on one of the key methods on Observable (`subscribe`).

To illustrate, consider this:

```java
    @Test
    public void shouldUnsubscribe() {
        TestSubscriber subscriber = new TestSubscriber();
        final Subscription subscription = Observable.just("item").subscribe(subscriber);
        subscription.unsubscribe();
        expect(subscriber.isUnsubscribed()).toBeTrue();
    }

```

This test passes. If I change line 2 of the test to this:
```java
        ...
        final Subscription subscription = Observable.just("item").subscribe((Observer) subscriber);
        ...
```

The test fails. The reason is the widening cast to `Observer`: this causes an overload of `subscribe` to be called which wraps away the observer into another Subscriber, without checking whether the observer isn't also a Subscriber and therefore has a Subscription that must be unsubscribed from. Since it uses the default constructor of Subscriber to perform the composition, no link will be established between the parent and child subscription, thus changing the behavior of the composition. 

I find this to be very surprising. The example is contrived, but this widening cast is actually very likely to happen whenever you're keeping a list of observers somewhere that can either be Subscribers or Observers. Doesn't this also mean that it will lead to very subtle bugs, because we're effectively disregarding the Subscription wrapped by the subscriber, which may want to call important cleanup logic in `unsubscribe`?

A simple but arguably not pretty fix for this would be to change the `subscribe` overload in a manner where it first tests whether the Observer is actually a Subscriber, and downcasts it safely, then only wrapping it in a Subscriber if necessary.

**Comment by akarnokd** [comment]
An ```instanceof``` test shoud do the trick.

**Comment by mttkay** [comment]
That sums it up nicely. :-)

**Comment by akarnokd** [comment]
Closing as the requested change was merged via #2630.
